openstates/me/people.py

```diff
@@ -102,20 +111,41 @@ def scrape_reps(self):
             member_name = " ".join((first_name, middle_name, last_name))
             member_name = re.sub(r'\s+', ' ', member_name).strip()
 
+            try:
+                # xlrd reads this value as a float. However we save it as a string.
+                district_number = "{:.0f}".format(float(d['district']))
+            except ValueError:
+                # Will reach here in case district number is blank. We will skip this row.
+                self.info("Skipping non voting member {}".format(member_name))
+                continue
+
             party = d['party']
             party = _party_map[party]
 
             # Determine legislator's URL to get their photo.
             # Roster URL list does not include nicknames, have to be removed before we search.
             if '"' in middle_name:
-                # Name contains nickname
-                roster_name = " ".join((first_name, ''.join(middle_name.split('"')[2:]),
-                                        last_name))
-                roster_name = re.sub(r'\s+', ' ', roster_name).strip()
+                # middle name contains nickname
+                roster_middle_name = ''.join(middle_name.split('"')[2:])
+                roster_middle_name = re.sub(r'\s+', ' ', roster_middle_name).strip()
             else:
-                roster_name = member_name
+                roster_middle_name = middle_name
+
+            # Searching for different parts of the name with an 'and' operator
+            # Since first_name middle_name last_name order not necessarily
+            # maintained in the roster
+            URL_XPATH = URL_XPATH = '//a[contains(text(), "Representative")'
+
+            # Get chunks of ascii characers in the name. lxml incorrectly decodes
+            # sometimes, hence skipping the non ascii parts of the names
+            for chunk in get_ascii_chunks(first_name):
+                URL_XPATH += ' and contains(text(), "{}")'.format(chunk)
+            for chunk in get_ascii_chunks(roster_middle_name):
+                URL_XPATH += ' and contains(text(), "{}")'.format(chunk)
+            for chunk in get_ascii_chunks(last_name):
+                URL_XPATH += ' and contains(text(), "{}")'.format(chunk)
+            URL_XPATH += ']/@href'
 
-            URL_XPATH = '//a[contains(text(), "{}")]/@href'.format(roster_name)
             try:
                 (leg_url, ) = roster_doc.xpath(URL_XPATH)
```

Comment:
Programmatically constructing XPath query strings isn't the best practice, and could be hard to read and debug. How about something very similar, but that does the filtering ex-post?

```
roster_name_chunks = get_ascii_chunks(first_name) + get_ascii_chunks(roster_middle_name) + get_ascii_chunks(last_name)
leg_url_elements = roster_doc.xpath('//a[contains(text(), "Representative")]')
leg_url = None
for el in leg_url_elements:
    if all(chunk in el.text_content() for chunk in roster_name_
chunks):
        # Avoid accidentally repeat-matching URLs
        assert leg_url is None
        leg_url = el.attrib['href']
if leg_url is None:
    self.warning("Could not find legislator URL for {}".format(member_name))
```

Reply:
I didn't 100% check that code, but the flow should be right, maybe look for syntax issues before using ;)